ARROW-13965: [C++] dynamic_casts in parquet TypedColumnWriterImpl impacting performance

etseidl opened the pull request:
This PR adds observer pointers to TypedColumnWriterImpl to avoid repeated dynamic_casts upon writing many small batches.  This can lead to significant performance gains.  As an example, using stream_reader_writer from examples/parquet/parquet_stream_api modified to write 20M rows, an improvement of over 30% was seen in write times.
```
original:
52.919u 2.146s 0:55.81 98.6%	0+0k 0+2646928io 0pf+0w
53.328u 2.327s 0:56.38 98.6%	0+0k 0+2646928io 0pf+0w
53.048u 2.341s 0:56.06 98.7%	0+0k 0+2646928io 0pf+0w

patched:
35.367u 2.012s 0:37.88 98.6%	0+0k 0+2646928io 0pf+0w
35.133u 1.717s 0:37.36 98.6%	0+0k 0+2646928io 0pf+0w
35.228u 1.950s 0:37.61 98.8%	0+0k 0+2646928io 0pf+0w
```

emkornfield commented on cpp/src/parquet/column_writer.cc:
```diff
@@ -1358,6 +1363,8 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
       current_encoder_ = MakeEncoder(DType::type_num, Encoding::PLAIN, false, descr_,
                                      properties_->memory_pool());
+      current_value_encoder_ = dynamic_cast<ValueEncoderType*>(current_encoder_.get());
+      current_dict_encoder_ = 0;  // not using dict
```
```suggestion
      current_dict_encoder_ = nullptr;  // not using dict
```

emkornfield reviewed (commented):
Is it possible to add a benchmark to show the improvement?  Or is there an existing one that shows this?

pitrou commented:
TBH, the stream APi is probably not high-performance at all. I'm not sure it makes sense to write benchmarks for it (unless we want to showcase how slow it is :-)).

etseidl commented:
Right now I'm trying to modify the stream_reader_writer example to see if the speed bump is noticeable there.  If not, I'll  modify my code (which doesn't actually use the streaming api, it just writes in batches of 1) to  do writes to a fixed schema.

etseidl commented:
Ok, I modified the stream_reader_writer example to just do the write, and changed the number of rows from 2000 to 20M (and the max row group size from 1000 to 10M).  I ran this on my laptop with and without the changes in this patch.  I used the csh time command to compare run times. The results are:
```
original:
52.919u 2.146s 0:55.81 98.6%	0+0k 0+2646928io 0pf+0w
53.328u 2.327s 0:56.38 98.6%	0+0k 0+2646928io 0pf+0w
53.048u 2.341s 0:56.06 98.7%	0+0k 0+2646928io 0pf+0w

patched:
35.367u 2.012s 0:37.88 98.6%	0+0k 0+2646928io 0pf+0w
35.133u 1.717s 0:37.36 98.6%	0+0k 0+2646928io 0pf+0w
35.228u 1.950s 0:37.61 98.8%	0+0k 0+2646928io 0pf+0w
```
@emkornfield Do you want benchmark code added to the PR, or are these results sufficient?

pitrou commented:
As far as am I concerned (and not wanting to speak for @emkornfield  :-)), I would be ok with simply adding these numbers to the PR description.

etseidl commented:
Sorry, I kept the PR description short since I had more details in the JIRA issue.  Once @emkornfield concurs, I'll edit the description.  Thanks!

emkornfield commented:
Yes, I think just updating the description is fine.

pitrou reviewed (approved):
Thank you @etseidl . I'll merge this if CI passes.

pitrou commented:
Note the LZ4 build failures on MinGW are unrelated.
